File: diet_guard/_kuchnia_client.py
"""HTTP access to the Kuchnia Wikinga panel.

Auth is a cookie session, not a bearer token: ``POST auth/login`` with a
**form-urlencoded** body sets a ``SESSION`` cookie that every later call
carries.  The cookie is cached (:mod:`diet_guard._kuchnia_config`) so a refresh
normally skips the login round trip entirely, and an expired cookie costs one
extra request rather than a failed import.

``requests`` is imported lazily.  The gate's not-due tick imports the CLI and
must not pay ~78ms for an HTTP stack it never touches -- the same reason
:mod:`diet_guard._estimator_off` and :mod:`diet_guard._cli_prune` defer theirs.
Call sites reach through ``sys.modules[__name__]`` so a test's ``patch.object``
still wins over the lazy hook.
"""

from __future__ import annotations

from importlib import import_module
import sys
import time
from typing import TYPE_CHECKING

from diet_guard._constants import (
    KUCHNIA_API_BASE,
    KUCHNIA_COMPANY,
    KUCHNIA_LAUNCHER_TYPE,
    KUCHNIA_TIMEOUT_SECONDS,
    KUCHNIA_TOTAL_DEADLINE_SECONDS,
)
from diet_guard._kuchnia_config import (
    SESSION_COOKIE,
    clear_session_cookie,
    load_session_cookie,
    read_credentials,
    save_session_cookie,
)
from diet_guard._kuchnia_errors import KuchniaError

if TYPE_CHECKING:  # pragma: no cover - typing only
    import requests

_LAZY_ATTRS = ("requests",)

_COOKIE_DOMAIN = "panel.kuchniavikinga.pl"


def __getattr__(name: str) -> object:
    """Resolve the deferred ``requests`` import on first attribute access."""
    if name not in _LAZY_ATTRS:
        msg = f"module {__name__!r} has no attribute {name!r}"
        raise AttributeError(msg)
    return import_module(name)


class PanelSession:
    """A logged-in panel session with a whole-walk deadline.

    The deadline exists because the import is several sequential requests: a
    per-request timeout alone would let a slow provider stack them into half a
    minute, and this runs on paths a user is waiting on.
    """

    def __init__(
        self, deadline_seconds: float = KUCHNIA_TOTAL_DEADLINE_SECONDS
    ) -> None:
        """Build an unauthenticated session with the panel's required headers."""
        module = sys.modules[__name__]
        self._session = module.requests.Session()
        # No session-level Content-Type: the login body is form-urlencoded and
        # a sticky JSON default would mislabel it. ``requests`` sets it per call.
        self._session.headers.update(
            {
                "company-id": KUCHNIA_COMPANY,
                "X-Launcher-Type": KUCHNIA_LAUNCHER_TYPE,
                "User-Agent": "diet_guard/1.0 (personal diet tracker)",
                "Accept": "application/json",
            },
        )
        self._expires_at = time.monotonic() + deadline_seconds

    def _check_deadline(self) -> None:
        """Raise once the whole-walk budget is spent."""
        if time.monotonic() >= self._expires_at:
            msg = "catering panel too slow (deadline exceeded)"
            raise KuchniaError(msg)

    def _request(
        self,
        method: str,
        path: str,
        data: dict[str, str] | None = None,
    ) -> requests.Response:
        """Perform one request, translating transport errors into KuchniaError.

        ``data`` is the only body this client ever sends (the form-urlencoded
        login), so it is spelled out rather than taken as ``**kwargs`` -- a
        concrete signature types better and keeps the surface honest.
        """
        self._check_deadline()
        module = sys.modules[__name__]
        try:
            return self._session.request(
                method,
                f"{KUCHNIA_API_BASE}/{path}",
                timeout=KUCHNIA_TIMEOUT_SECONDS,
                data=data,
            )
        except module.requests.RequestException as exc:
            msg = f"catering panel unreachable: {exc}"
            raise KuchniaError(msg) from exc

    def _login(self) -> None:
        """Authenticate with the stored credentials and cache the cookie."""
        username, password = read_credentials()
        response = self._request(
            "POST",
            "auth/login",
            data={"username": username, "password": password},
        )
        if not response.ok:
            msg = f"catering login rejected (HTTP {response.status_code})"
            raise KuchniaError(msg)
        cookie = self._session.cookies.get(SESSION_COOKIE)
        if cookie is None:
            msg = "catering login returned no session cookie"
            raise KuchniaError(msg)
        save_session_cookie(cookie)

    def authenticate(self) -> None:
        """Attach the cached cookie, logging in only when there is not one."""
        cached = load_session_cookie()
        if cached is None:
            self._login()
            return
        self._session.cookies.set(SESSION_COOKIE, cached, domain=_COOKIE_DOMAIN)

    def get_json(self, path: str) -> object:
        """GET ``path`` and decode it, re-logging in once if the cookie expired.

        A cached cookie that has expired is indistinguishable from a good one
        until it is used, so an auth failure is retried exactly once with fresh
        credentials. A second failure is real.

        Args:
            path: Path below the API base, without a leading slash.

        Returns:
            The decoded JSON body.

        Raises:
            KuchniaError: On a non-OK status or an undecodable body.
        """
        response = self._request("GET", path)
        if response.status_code in {401, 403}:
            clear_session_cookie()
            self._login()
            response = self._request("GET", path)
        if not response.ok:
            msg = f"catering panel returned HTTP {response.status_code} for {path}"
            raise KuchniaError(msg)
        try:
            return response.json()
        except ValueError as exc:
            msg = f"catering panel returned a non-JSON body for {path}"
            raise KuchniaError(msg) from exc

File: app/lib/services/kuchnia_client.dart
/// HTTP access to the Kuchnia Wikinga panel, from the phone.
///
/// The Dart mirror of `diet_guard/_kuchnia_client.py`. Auth is a cookie
/// session, not a bearer token: `POST auth/login` with a **form-urlencoded**
/// body sets a `SESSION` cookie that every later call carries.
///
/// `package:http` has no cookie jar, so the `set-cookie` header is read and
/// replayed by hand. `Response.headers` folds duplicate `set-cookie` values
/// into one comma-joined string, which is why the parser below splits and
/// looks for the named cookie rather than assuming the header holds only it.
///
/// **Android only.** The panel sends no CORS headers, so a browser cannot call
/// it at all; on web every entry point returns an unsupported reason rather
/// than throwing. The `kIsWeb` check comes first, before anything
/// platform-shaped, and this file imports no `dart:io` -- `package:http`
/// compiles for web, it is the *panel* that refuses, not the toolchain.
library;

import 'dart:convert';

import 'package:diet_guard_app/services/kuchnia_errors.dart';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;

/// Base URL of the panel's undocumented API.
const kuchniaApiBase = 'https://panel.kuchniavikinga.pl/api';

/// Sent as the `company-id` header. The panel wants the company *name*, not a
/// numeric id.
const kuchniaCompany = 'kuchniavikinga';

/// The panel's own launcher tag for a browser panel session.
const kuchniaLauncherType = 'BROWSER_PANEL';

/// The only cookie the panel issues on login.
///
/// It sets no `XSRF-TOKEN`, so the CSRF echo its own JavaScript performs is a
/// no-op for this account -- the session cookie alone authenticates.
const kuchniaSessionCookie = 'SESSION';

/// Per-request timeout.
const kuchniaTimeout = Duration(seconds: 8);

/// Whole-walk ceiling.
///
/// The import is three or four sequential requests, so a per-request timeout
/// alone would permit ~32s against a slow provider -- and this runs on a path
/// the user is waiting on.
const kuchniaTotalDeadline = Duration(seconds: 12);

/// Why the catering panel cannot be reached from a browser.
const kuchniaWebUnsupportedReason =
    'Android only — the catering panel blocks browser requests';

/// Extracts [name]'s value from a (possibly comma-joined) `set-cookie` header.
///
/// `http`'s `Response.headers` collapses repeated headers into one string, so
/// a panel that sets more than one cookie hands over
/// `SESSION=abc; Path=/, OTHER=def; Path=/`. Returns null when the cookie is
/// absent, which the caller treats as a failed login rather than guessing.
String? sessionCookieFrom(String? setCookieHeader) {
  if (setCookieHeader == null || setCookieHeader.isEmpty) return null;
  // Split on both separators: ',' joins whole cookies, ';' separates a
  // cookie's attributes from its value.
  for (final part in setCookieHeader.split(RegExp('[,;]'))) {
    final trimmed = part.trim();
    if (trimmed.startsWith('$kuchniaSessionCookie=')) {
      final value = trimmed.substring(kuchniaSessionCookie.length + 1);
      if (value.isNotEmpty) return value;
    }
  }
  return null;
}

/// A logged-in panel session with a whole-walk deadline.
class KuchniaSession {
  /// Creates a session over [client], authenticating lazily.
  KuchniaSession({
    required this.username,
    required this.password,
    http.Client? client,
    DateTime? startedAt,
  }) : _client = client ?? http.Client(),
       _deadline =
           (startedAt ?? DateTime.now()).add(kuchniaTotalDeadline);

  /// The panel e-mail.
  final String username;

  /// The panel password.
  final String password;

  final http.Client _client;
  final DateTime _deadline;

  /// The session cookie: seeded from a cache to skip the login, and read back
  /// afterwards so the caller can persist a freshly minted one.
  ///
  /// Device-local by design -- it is regenerable from the password, so syncing
  /// it would widen exposure and buy nothing.
  String? sessionCookie;

  Map<String, String> get _headers => {
    'company-id': kuchniaCompany,
    'X-Launcher-Type': kuchniaLauncherType,
    'User-Agent': 'diet_guard/1.0 (personal diet tracker)',
    'Accept': 'application/json',
    if (sessionCookie != null) 'Cookie': '$kuchniaSessionCookie=$sessionCookie',
  };

  void _checkDeadline() {
    if (!DateTime.now().isBefore(_deadline)) {
      throw const KuchniaError('catering panel too slow (deadline exceeded)');
    }
  }

  /// Authenticates and caches the session cookie.
  ///
  /// No client-wide `Content-Type`: the login body is form-urlencoded and a
  /// sticky JSON default would mislabel it. Passing a `Map` to `http.post`
  /// sets the right one automatically.
  Future<void> login() async {
    _checkDeadline();
    final http.Response response;
    try {
      response = await _client
          .post(
            Uri.parse('$kuchniaApiBase/auth/login'),
            headers: _headers,
            body: {'username': username, 'password': password},
          )
          .timeout(kuchniaTimeout);
    } on Exception catch (error) {
      throw KuchniaError('catering panel unreachable: $error');
    }
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw KuchniaError(
        'catering login rejected (HTTP ${response.statusCode})',
      );
    }
    final cookie = sessionCookieFrom(response.headers['set-cookie']);
    if (cookie == null) {
      throw const KuchniaError('catering login returned no session cookie');
    }
    sessionCookie = cookie;
  }

  /// GETs [path] and decodes it, re-logging in once if the cookie expired.
  ///
  /// A cached cookie that has expired is indistinguishable from a good one
  /// until it is used, so an auth failure is retried exactly once with fresh
  /// credentials. A second failure is real.
  Future<Object?> getJson(String path) async {
    if (sessionCookie == null) await login();
    var response = await _get(path);
    if (response.statusCode == 401 || response.statusCode == 403) {
      sessionCookie = null;
      await login();
      response = await _get(path);
    }
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw KuchniaError(
        'catering panel returned HTTP ${response.statusCode} for $path',
      );
    }
    try {
      return jsonDecode(response.body);
    } on FormatException {
      throw KuchniaError('catering panel returned a non-JSON body for $path');
    }
  }

  Future<http.Response> _get(String path) async {
    _checkDeadline();
    try {
      return await _client
          .get(Uri.parse('$kuchniaApiBase/$path'), headers: _headers)
          .timeout(kuchniaTimeout);
    } on Exception catch (error) {
      throw KuchniaError('catering panel unreachable: $error');
    }
  }

  /// Releases the underlying HTTP client.
  void close() => _client.close();
}

/// True when this platform can reach the catering panel at all.
///
/// False on web: the panel sends no CORS headers, so the browser refuses the
/// request before it leaves. Checked before anything platform-shaped runs.
bool get kuchniaFetchSupported => !kIsWeb;
